### PR TITLE
Add message listening system

### PR DESCRIPTION
**Pull request by Patrick-van-Halm**
Adds the system to listen to Frida agent messages. Requires pull request: [Publicize setMethodCallHandler](https://github.com/httptoolkit/dbus-native/pull/1)

Resolves: #3

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -98,6 +104,16 @@ export class FridaSession {
             .getInterface<AgentSession>('/re/frida/AgentSession/' + sessionId, 're.frida.AgentSession16');
     }
 
+    private listenToAgentMessages(sessionId: string, cb: (message: ScriptAgentMessage) => void){
+        this.bus.setMethodCallHandler(`/re/frida/AgentMessageSink/${sessionId}`, "re.frida.AgentMessageSink16", "PostMessages", [(messages: AgentMessage[]) => {
+            for(const message of messages) { // ScriptMessage 
+                if(message[0] === 1) {
+                    return cb(JSON.parse(message[2]));
```
Comment (pimterry) [inline src/index.ts:111]: I'm not totally clear how the dbus message format works here, what the `=== 1` check is doing, or how we know `[2]` is always a message with `type: "send"`.

Is it possible to add a comment linking some docs for this format, or the corresponding Frida code defining it or similar?

A constant for the 1 that gives a hint at what it means (and what any non-1 values might mean) would also be helpful.

Comment (pimterry) [inline src/index.ts:111]: This `return` means we'll always only listen to the first message with the `message[0] === 1` in the batch. Is that correct?

File: test/test.spec.ts [hunk 2]
```diff
@@ -223,6 +223,67 @@ describe("Frida-JS", () => {
...
+                const interval = setInterval(() => {
+                    if (message) {
+                        resolve();
+                    }
+                }, 100);
+                setTimeout(() => {
+                    clearInterval(interval);
+                    reject(new Error('Timed out waiting for message'));
+                }, 5000);
+            });
```
Comment (pimterry) [inline test/test.spec.ts:281]: You can simplify this and remove the interval by moving the `listenToSession` code inside the promise, and calling resolve there directly with the message when it's received:

```javascript
const message = await new Promise<ScriptAgentMessage>((resolve, reject) => {
    fridaClient.listenToSession(id, resolve);
    setTimeout(() => {
        reject(new Error('Timed out waiting for message'));
    }, 5000);
});
```

**Review by pimterry — CHANGES_REQUESTED** [review]
This is awesome! Thanks so much for looking into this, I've been meaning to investigate this for a while so this is super helpful.

A few comments here about the details but in general this looks great.

File: src/index.ts [hunk 3]
```diff
@@ -157,6 +173,8 @@ export class FridaSession {
 
         const scriptId = await agentSession.CreateScript(fridaScript, {});
         await agentSession.LoadScript(scriptId);
+
+        return sessionId;
```
[the remark this answers is not in the document]
Comment (Patrick-van-Halm) [reply src/index.ts:177]: Yeah that sounds good. I wanted to do it like this but was a little hesitant with changing the code base too much haha

File: src/index.ts (hunk 1 above)
Comment (Patrick-van-Halm) [reply src/index.ts:111]: Ahh good one, I overlooked it myself last night. It should've been a continue statement.

**Comment by pimterry** [comment]
I've now published dbus-native too, so you can update that here whenever you're ready.

By the way, in case you're not aware, [HTTP Toolkit Pro](https://httptoolkit.com/get-pro/) is totally free for all contributors to the every repo in this org, and this is already a very helpful contribution already! I've just set you up an account linked to your git email `pv...@gmail.com`. If that sounds useful, you can install HTTP Toolkit from [httptoolkit.com](https://httptoolkit.com), click 'Get Pro' then 'Log into existing account' and enter your email to use it.

**Comment by Patrick-van-Halm** [comment]
> and this is already a very helpful contribution already! I've just set you up an account linked to your git email `pv...@gmail.com`. If that sounds

Thank you very much I will definitely look into it :)

File: package.json [hunk 4]
```diff
@@ -10,7 +10,7 @@
     "scripts/"
   ],
   "scripts": {
-    "prebuild": "rimraf dist/*",
+    "prebuild": "rimraf dist/",
```
Comment (Patrick-van-Halm) [inline package.json:13]: If you want I can revert this, but on windows it threw an error.

**Comment by Patrick-van-Halm** [comment]
@pimterry Interesting how the CI fails on a different test every time. Could you take a look at the code? I applied the requested changes.

**Comment by pimterry** [comment]
Yeah, I'm not sure what's happening there. There's definitely been some occasional flakiness in the past, but this seems very unusual. I'm going to do some more testing and see if I can get that fixed.

File: src/index.ts [hunk 5]
```diff
@@ -73,6 +73,44 @@ interface AgentSession {
...
+ * https://github.com/frida/frida-node/blob/main/lib/script.ts#L103-L115
+ */
+export enum MessageType {
+    Send = "send",
+    Error = "error"
+}
+
+export type Message = ScriptAgentSendMessage | ScriptAgentErrorMessage;
+export type ScriptAgentSendMessage = {
+    type: MessageType,
```
Comment (pimterry) [inline src/index.ts:98]: I think this should be `MessageType.Send`

File: test/run-frida-server.ts [hunk 6]
```diff
@@ -7,7 +7,7 @@ import { delay } from './test-util';
 let fridaServer: ChildProcess | undefined;
 
 export const FRIDA_SERVER_DIR = path.join(__dirname, '.frida-server');
-export const FRIDA_SERVER_BIN = path.join(FRIDA_SERVER_DIR, 'frida-server');
+export const FRIDA_SERVER_BIN = path.join(FRIDA_SERVER_DIR, 'frida-server.exe');
```
Comment (pimterry) [inline test/run-frida-server.ts:10]: I think this should be platform specific, just `.exe` for Windows

**Review by pimterry — CHANGES_REQUESTED** [review]
I've now fixed the tests on main and the seem to be much more reliable. I've rebased the PR and it's all passing :+1:

Thanks for those quick changes on this @Patrick-van-Halm! I've added a last couple of tiny comments here, but then I think this will be ready to merge.

**Review by pimterry — APPROVED** [review]

**Comment by pimterry** [comment]
Awesome! Thanks for all the hard work here, this is great :+1:. I'll do a release in just a sec.

**Pull request merged**